Enhancement for enabling single-node-openshift with workers (#1041)
openshift/enhancements: enhancements/single-node/single-node-openshift-with-workers.md

with additional workers recovers after the single control-plane node reboots by
running conformance tests post-reboot.

- Add tests on both cloud / "none"-platform that check that a single-node cluster
with additional workers recovers after an upgrade by running conformance tests
post-upgrade.

TODO: Describe day-1-workers tests?

### Graduation Criteria

**Note:** *Section not required until targeted at a release.*

TBD

#### Dev Preview -> Tech Preview

TBD

#### Tech Preview -> GA

TBD

#### Removing a deprecated feature

N/A

### Upgrade / Downgrade Strategy

In the non-goals section it's mentioned that this enhancement does not apply to
clusters which have been installed prior to the enhancement, so their upgrade
is not discussed.

This enhancement, to the best of my knowledge, should have no problems
persisting across cluster upgrades. The Test Plan section describes how this
will be tested.

### Version Skew Strategy

Does not apply, to the best of my understanding.

### Operational Aspects of API Extensions

- The "ControlPlane" IngressPlacement parameter value may not be used if the cluster's

> ControlPlaneTopology parameter is set to "External". 

Reviewer: Specifying `ingressPlacement: ControlPlane` with `controlPlaneTopology: External` is an interesting combination.  Maybe we can leave that as reserved for possible future use.

Author: Not sure I understand what you mean by "Leaving it as reserved"? For now I'll say that this combination is forbidden, I don't think it will be a "Breaking change" to say otherwise in the future

Reviewer: Yeah, that's what I mean.  It might be useful for, say, some Hypershift use-case, so I'd rather say it is invalid now but not necessarily invalid forever.

> - This API change only affects the defaulting behavior of the IngressController CR,
> it does not add any new capabilities to OCP, or gives any more flexibility than there
> already was.

Reviewer: ```suggestion
- This API change only affects the defaulting behavior of the IngressController CR,
it does not add any new capabilities to OCP, or give any more flexibility than there
already was.
```

- For administrators and support engineers, the IngressController is still the source
of truth and where you need to look for understanding the router placement in practice.
Nothing has changed in that regard.

#### Failure Modes

- The "ControlPlane" IngressPlacement parameter value may not be used if the cluster's
ControlPlaneTopology parameter is set to "External". This should ideally be blocked by
admission controllers for the Ingress config CR, if there are any. On top of that, the
Ingress Operator should refuse to create an IngressController CR or reconcile an existing
one while IngressPlacement is set to "ControlPlane" and the ControlPlaneTopology parameter

> is set to "External". This will harm cluster health and disable ingress traffic to the cluster.

Reviewer: Can the installer's validation reject that combination?  I don't know if there's an admission plugin that prevents the user from updating these fields post-install.  

I don't think the ingress operator should refuse to create a default ingresscontroller in this scenario; the operator should try its hardest to make the cluster functional.  How about if the operator simply ignore `ingressPlacement` if `controlPlaneTopology` is "External"?

Reviewer 2: The installer will not create an Infrastructure CR with the `controlPlaneTopology` set to "External". However, the installer does not validate modifications made by the user to manifests. There is the possibility that user (incorrectly) sets the `controlPlaneTopology`.

Author: >Can the installer's validation reject that combination?

So following what @staebler said, I guess the answer to that is no

>I don't know if there's an admission plugin that prevents the user from updating these fields post-install.

OK

>I don't think the ingress operator should refuse to create a default ingresscontroller in this scenario; the operator should try its hardest to make the cluster functional. How about if the operator simply ignore ingressPlacement if controlPlaneTopology is "External"?

Fair, I wasn't sure what's the attitude towards misconfigurations. So you're saying the behavior should be completely ignoring it (i.e., treating it as if it were Workers)?

Should the ingress operator log a warning? Set some .status on some CR? Or just flat out ignore it?

Reviewer: And yes, I mean treat `ControlPlane`/`External` as if `ingressPlacement` were set to `Workers`.

may no longer work. This can be temporarily solved if DNS is adjusted to point
at the correct node currently holding the pod or a load-balancer / some virtual
IP solution is put in place and then the DNS entry can be directed at that
load-balancer / virtual IP instead of at the node. This enhancement will try
to prevent this floating altogether so these workarounds will not be needed.

Finally, performing bootstrap-in-place installation of a "none"-platform single
control-plane node cluster today using the OpenShift Installer does not provide
the user with a worker node ignition file. As a result, the user is not able to
manually add worker nodes to the resulting cluster. All current official OCP
documentation points the user to use "the worker ignition file that get
generated during installation", but that is not currently generated in such
bootstrap-in-place installation.

### Goals

- Deal with the "floating ingress" issue encountered in "none"-platform
single control-plane node clusters which have worker nodes added to them.

- Define how the installer may be modified to generate the worker ignition
manifest, even when doing bootstrap-in-place "none"-platform single

> control-plane node cluster installation

Reviewer 3: I didn't see this fleshed out anywhere, have I missed it?

Author: Ah yes, sorry, used to have a TBD for this, looks like it accidentally got deleted during my refactoring work for the enhancement.

Anyway, the installer explicitly avoids generating the `worker.ign` files for BIP installations, this can be trivially fixed by telling it to stop doing that (that == avoiding), I just haven't found the exact piece of code that does that so I didn't mention it in the enhancement, but it should be a trivial matter.

schedulable, and as a result a lot of infrastructure workloads fail to schedule
and operators degrade. User intervention is required to fix this.

- Deal with expansion of the single-node control-plane by adding more
control-plane nodes

- Deal with expansion of clusters that have been installed before the
implementation of this enhancement (if possible, their expansion may be
addressed by an upgrade followed with documentation outlining the steps that
the user has to take to enable expansion, e.g. patch the default
`IngressController` to have node selector targetting the master pool rather
than the worker pool).

- Adjust the baremetal platform to support single control-plane node
installations. The baremetal platform solves the "floating ingress" issue by
using virtual IP addresses/keepalived.

- Deal with users who want their ingress traffic to not go through the single
control-plane node.

- Deal with non-cloud, non-"none" platforms such as baremetal, vSphere, etc.

## Proposal

- Create a new `IngressPlacement` API parameter.

- Make sure worker node ignition files are generated even in bootstrap-in-place

> single control-plane node "none"-platform installation.

Reviewer 3: Where is this fleshed out in the enhancement?

Author: https://github.com/openshift/enhancements/pull/1041#discussion_r820631305

### User Stories

- As an OpenShift cluster administrator, I want to install a single control-plane
node cluster with additional worker nodes, so that my cluster will be able to
handle larger computation demands.

- As an OpenShift cluster administrator, I want to add worker nodes to my
existing single control-plane node cluster, so that it'll be able to meet
growing computation demands.

### API Extensions

> Introduce a new topology field in the Ingress config CR

Reviewer 4: ```suggestion
Introduce a new field in the Ingress config CR
```

> (`config.openshift.io/v1/ingresses`) called `IngressPlacement`.

Reviewer 3: Is it possible we could flesh out the API here in Go format to match what will be added to the API, eg

```
type IngressSpec struct {
  // Existing fields omitted

  // ingressPlacement ...
  // +kubebuilder:validation:Enum:="ControlPlane";"Worker"
  // +kubebuidler:default:="Worker"
  // +optional
  IngressPlacement string `json:"ingressPlacement"`
}
```

It would also be good to think about a few things here:
- What does the field do and how will it be used by users
- Given the existing possible values mentioned below, do we expect that there may ever be any other configuration to the placement? Eg if this were a struct it could have a field for `role` which would select on the role of the Machine, and then maybe some other fields in the future as needed?
- Given this field is on the `Ingress` object, does it need to be `IngressPlacement` or should it just be `Placement`?
- We are making an effort to ensure appropriate kubebuilder validation is present on all fields going into the API, so it would be good to represent that here, is it optional/required for example, if it's an Enum, then add an Enum tag or make it a typed string which will generate the same OpenAPI validation

Author: Yeah I started work on doing something similar in the openshift/api repo. Since your comments lowered my confidence in this solution, I'll wait for us to agree before making this change to the enhancement

### Implementation Details/Notes/Constraints

> This new field will have one of these values - `ControlPlane` or `Workers`. 

Reviewer 3: `Workers` here seems wrong, what's the motivation for `Workers` being plural rather than just `Worker`, everywhere else we use a singular (for example in role labels, node selectors)

Author: "Place the Ingress on the ControlPlane nodes or on the Worker node(s) AKA Workers" is what I had in mind. I feel Worker is a bit misleading, sounds like the ingress is gonna be on one worker, which is not necessarily true

In addition, continue to allow the `.spec.replicas` and `.spec.nodePlacement`
parameters in `operator.openshift.io/v1/ingresscontrollers` CRs to be omitted,

> but change the defaulting behavior for these fields.

Reviewer 3: I don't think changing the defaulting behaviour is something that we can do. Typically this would be considered a breaking change as existing consumers may be relying on the existing defaults. Is this considered to be a workload API or a configuration API? If the number of users of the API is sufficiently small then changing the defaults can be possible but it needs to be considered the impact of the change of this default

Edit: Having read further, it's unclear to me what we mean by defaulting, it would be useful to clarify if we are updating the spec of the object during the defaulting or whether this is controller side defaulting only

Author: >Is this considered to be a workload API or a configuration API? If the number of users of the API is sufficiently small then changing the defaults can be possible but it needs to be considered the impact of the change of this default

This approach was inspired by @Miciah's  openshift/api#1103 - I guess defaults are an API guarantee and this is truly a breaking change, but would that mean that the linked PR is also a breaking change?

I personally took "optional" omitted parameters in a CR to mean the user saying "I don't have an opinion on this, do whatever makes sense" and if you look at it that way it's not really a breaking change... I'm torn on that.

What if we instead limited the scope of this new API field to affect the defaulting behavior of just the router-default IngressController ? 

>Edit: Having read further, it's unclear to me what we mean by defaulting, it would be useful to clarify if we are updating the spec of the object during the defaulting or whether this is controller side defaulting only

It's definitely a controller thing - the controller creates the initial IngressController if there isn't one already, and its behavior while doing so according to the new API. And it should also initialize those parameters if the installer creates an IngressController with those parameters omitted

The value of the `IngressPlacement` field will affect the defaulting behavior
of the `IngressController`'s `.spec.replicas` and `.spec.nodePlacement`
parameters.  In the absence of an `IngressController` resource created by the
user/installer, or when the user/installer creates an `IngressController` with

> these two parameters omitted, the Cluster Ingress Operator will choose the

Reviewer 3: Do you mean that the values will be set on the object after create?

Have you considered a defaulting webhook for this rather than a controller? Using a controller implies that the change will be asynchronous. This means a user could create the object and observe their desired configuration without the defaults, then a moment later observe it with the results. Defaulting is better handled synchronously to avoid this discrepancy on the consumer side

Author: >Have you considered a defaulting webhook for this rather than a controller?

Since the router-default IngressController may be created by the Installer as an installer manifest, do "defaulting webhooks" this early in the cluster's lifecycle even exist / make sense?

default values for those parameters based on the value of `IngressPlacement`.

If the value of `IngressPlacement` itself is omitted by the installer, it
is defaulted to `Workers`. This would be useful to maintain the current behavior
if the API change is merged before the installer change to set this field, or if
we want this newer OCP version to be backwards-compatible with older installers.

When the value of `IngressPlacement` is `Workers`, the defaulting behavior of
`.spec.replicas` and `.spec.nodePlacement` will be the same as it is today:
`.spec.replicas` will be chosen according to the value of
`InfrastructureTopology`, namely `1` when `SingleReplica` or `2` when

> `HighlyAvailable`. `.spec.nodePlacement` will always just be:

Reviewer 3: These fields are currently optional, if you were thinking to have a controller/something else set them, then perhaps it is better not to do so. My suggestion would be to update the descriptions of these fields to have some wording along the lines of
```
When this field is omitted, it is up to the platform to decide a reasonable default. This value will be based upon ... and may change over time.
```

Author: Right

```yaml
nodePlacement:
  nodeSelector:
    matchLabels:
      kubernetes.io/os: linux
      node-role.kubernetes.io/worker: ''
```

However, if the value of `IngressPlacement` is `ControlPlane`, the defaulting
behavior will be different: `.spec.replicas` will be chosen instead according
to the value of `ControlPlaneTopology`; again, `1` when `SingleReplica` or `2`
when `HighlyAvailable`. `.spec.nodePlacement` will be always just be:

```yaml
nodePlacement:
  nodeSelector:
    matchLabels:
      kubernetes.io/os: linux
      node-role.kubernetes.io/master: ''
```

(Note that the `kubernetes.io/os: linux` label is mentioned just because it's
the current behavior, it has no importance in this enhancement)

> The installer will detect situations in which it's unlikely the user will want

Reviewer 4: @patrickdillon @staebler are you OK with having the installer include this logic?

Author: Ongoing discussion at https://github.com/openshift/enhancements/pull/1041#discussion_r817700777

to set up a load-balancer. For now, those situations only include installation
of single control-plane node cluster deployments on the "none" platform. In
those situations, the installer will set `IngressPlacement` to be
`ControlPlane`. Since there's just a single control-plane node, `ControlPlane`
topology would be `SingleReplica` and the combined effect would be that the
`IngressController` will have just a single replica and be pinned to the single
control-plane node. This will then ensure that the `router-default` deployment
created by the Cluster Ingress Operator will always run on the single
control-plane node, and as a result any `*.apps.<cluster>.<base>` DNS entries
which originally pointed at the single control-plane node will remain correct
even in the face of newly added worker nodes.

In any other situations, the installer will set `IngressPlacement` to
`Workers`, resulting in the same default behavior as before this enhancement,
namely that `IngressController` pod replicas are scheduled on worker nodes and
determined according to the `InfrastructureTopology`.

> 

Reviewer 4: Can the user influence the setting through the installer by asking for it to be set explicitly?

Can the user change the setting in a running cluster?

Author: >Can the user influence the setting through the installer by asking for it to be set explicitly?

No plans for that at the moment, but there's nothing stopping us from adding such configuration to the installer in the future, if we want users to be exposed to this

>Can the user change the setting in a running cluster?

I'm not sure about the mutability of the Ingress config CR in general. Can our API declare whether specific CR parameters are configurable, or does mutability apply to the entire CR as a whole?

I don't see a reason why users may even want to change this - users can just modify the IngressController directly to match the configuration they desire. There's no reason for them to edit a configuration whose sole purpose is to affect the defaulting behavior of new (or the first router-default) IngressController(s).

If we do make this parameter mutable, changing it should have no effect on existing IngressController CRs, since the effect of this parameter is just how IngressControllers get their parameters set by default, and since they already have been created, they already have those parameters set, so there's nothing to default.

Given this kind of unintuitive "this parameter only controls the defaulting behavior of another CR" I lean towards leaving it immutable to avoid the confusion and not give users the false impression that changing it after the cluster is up and running would have any effect on the existing IngressControllers.

Let me know what you think, and I'll write that down more concretely in the enhancement.

Reviewer 3: > Can our API declare whether specific CR parameters are configurable

In general we should assume that any spec field is mutable. If you need immutability, this can only be achieved via a validating webhook at the moment

> I don't see a reason why users may even want to change this

It's important to note in the enhancement the impact of what would happen if they did, because the likelihood is that some user will try

Reading through this comment you've assumed fields are immutable, which has influenced your reply, is there anything you would change about your comment given all fields are mutable by default?

Author: >In general we should assume that any spec field is mutable. If you need immutability, this can only be achieved via a validating webhook at the moment

Good to know. I was told that we don't currently have a validating webhook for this resource, and I'm not sure this would be a good enough reason to add one.

>Reading through this comment you've assumed fields are immutable, which has influenced your reply, is there anything you would change about your comment given all fields are mutable by default?

I have not made that assumption, it doesn't have to be immutable, it's just my preference that it will be, because of its confusing natuve - I wouldn't want users to touch it.

But if it can't be trivially made immutable, then so be it - It would have to stay mutable and confusing.

Or alternatively, if we can come up with a way to solve our problems without this new API, then we wouldn't need it at all.

Reviewer 3: > I was told that we don't currently have a validating webhook for this resource, and I'm not sure this would be a good enough reason to add one.

I suspect something like this has been said many times before, if every enhancement wants one small piece of validation, but none warrants it alone, it will never be done, but if you look wider, it would probably be worth it to start it, and then other validations may follow

> I have not made that assumption, it doesn't have to be immutable, it's just my preference that it will be, because of its confusing nature - I wouldn't want users to touch it.

Sorry, I was posing my point off of "If we do make this parameter mutable", I think if we believe the parameter is confusing, or should be immutable, there are other ways that we can do this. For example, if you were to add the parameter to the status of some object (I guess the IngressController is created by the installer and is a singleton right?), then it is _harder_ for a user to modify it. Generally spec fields are considered to be owned by the user, where status is owned by the cluster. We have used status fields on objects (eg Infrastructure) before for information that we want to have set from install time that we then prefer to be immutable later

Reviewer 4: In general, we can't just tell users not to touch fields. If they can modify it, we need to anticipate that they will. If we want the fields to be immutable, we need to do the work to make them immutable.

The existing topology fields are good examples of fields that we want to consider immutable right now, but that might be mutable in the future. Those are in the status field of the infrastructure CR.

Reviewer 3: Just a note on this, I'm ok with us putting this in the status to make it "immutable", but we've done this in the past for other things that we have then moved into the spec, now we have a deprecated field in the status which must be observed forever. My concern is we are creating future toil here by trying to make this immutable

I would suggest that you consider whether we can just allow the user to change the `DefaultPlacement` field value, I think given the description on the field (which is now really thorough, thank you!), we can probably say that we have documented the behaviour of the field sufficiently such that a user should understand that if it's changed, it won't affect existing `IngressController` instances. Were there any other concerns about this field being mutable?

In the future, when IPI-installed single control-plane node clusters in the
cloud no longer provision a load-balancer by default, they would also benefit

> from having the installer set the `IngressPlacement` to `ControlPlane`.

Reviewer 3: Is this a thing that is happening? I wasn't aware

### Risks and Mitigations

This should make no noticable difference on "regular" single control-plane node
clusters which do not have any day-1 or day-2 worker nodes. The only difference
for those clusters would be the `IngressController` targeting the "master" pool
rather than the "worker" pool, but since the single control-plane node is
already both in the "master" and "worker" pools, that should make no practical

> difference.

Reviewer 4: Is there any impact on multi-node clusters? The default behavior won't change, but is there any reason to be concerned if the user changes the setting to "ControlPlane" themselves? Can they break the cluster?

Author: I think for now we're not gonna let the user set this parameter, so there's no risk of it being applied in non-single-node clusters.

Having said that, hypothetically if the user was able to set this new parameter to ControlPlane on non-single-node clusters, especially ones that have installer-provisioned load-balancer. I think it does have the *potential* to mess things up - it would depend heavily on how the load balancer is being set up to find out which nodes it should target.

I don't know the details in reality, so maybe it's not an actual problem, but I imagine this hypothetical scenario -

- User chooses 3 control-plane 3 worker cluster installation on <some potentially problematic platform>
- User chooses IngressPlacement defaulting behavior to target ControlPlane
- Platform's installer naively points the provisioned load balancer only at the worker nodes (since it sees that this is not a compact cluster where control-plane nodes are also considered workers), that platform's installer does that because it's completely unaware of the effects of the IngressPlacement parameter* 
- IngressController gets created with 2 replicas (because master topology is HA) and it would node-select the master pool, 2 router-default pod gets placed on 2 random control-plane nodes
- Ingress traffic is broken because the load balancer is not pointing at the correct nodes

The bullet marked with * at the end of it is where I have a bit of uncertainty - are any platform installer's doing this "naive" action of pointing the load balancer only at the cluster's worker nodes? If yes, this is gonna be a problem that's gonna have to be solved whenever we make this parameter user-configurable

Reviewer 3: > I think for now we're not gonna let the user set this parameter, so there's no risk of it being applied in non-single-node clusters.

How will you enforce this?

Author: >How will you enforce this?

See https://github.com/openshift/enhancements/pull/1041#discussion_r820636797

Reviewer: > The installer does not create the load balancer that ingress uses.

It does for at least one platform, namely Azure (see [BZ#1794839](https://bugzilla.redhat.com/show_bug.cgi?id=1794839)).  However, in the case of Azure, the load balancer always includes both control-plane as well as worker nodes.

Reviewer 2: Yes, you are technically correct that, in the case of Azure, the installer creates the load balancer used by ingress. However, the installer does not create the backends, which is the part of the load balancer that is germane to this discussion.

any problems not described in this enhancement.

- Add periodic nightly tests which install a single-node "none"-platform cluster,
add worker nodes to it, and check that ingress traffic still works as expected
and recovers even after the `router-default` pod gets deleted and rescheduled.
Make sure this is still true even after upgrades.

- Add tests on both cloud / "none"-platform that check that a single-node cluster
with additional workers recovers after the single control-plane node reboots by
running conformance tests post-reboot.

- Add tests on both cloud / "none"-platform that check that a single-node cluster
with additional workers recovers after an upgrade by running conformance tests
post-upgrade.

TODO: Describe day-1-workers tests?

### Graduation Criteria

**Note:** *Section not required until targeted at a release.*

TBD

#### Dev Preview -> Tech Preview

TBD

#### Tech Preview -> GA

TBD

#### Removing a deprecated feature

N/A

### Upgrade / Downgrade Strategy

In the non-goals section it's mentioned that this enhancement does not apply to
clusters which have been installed prior to the enhancement, so their upgrade

> is not discussed.

Reviewer 4: I don't think we're going to be able to use that as a reason not to talk about upgrades here, because this change doesn't only apply to SNO deployments. All existing HA deployments will have the new field after an upgrade.

Author: Since these platforms already have IngressController CRs defined, this parameter should have no effect on them, as it only controls the defaulting behavior of such resources, and they already have those parameters set. 

You are correct that it has to be addressed. I will make sure it is

Author: EDIT: It seems that `replicas` is set explicitly, but the node selector is not. I'll give that some thought, if it could maybe cause problems

Reviewer: Correct, the IngressController CR can be deleted.  If the user deletes the default IngressController CR, then the ingress operator recreates it after a short delay, and at that point, this enhancement would influence the defaults for the newly created CR.  That wouldn't happen during an upgrade or downgrade though, unless the user were deleting the CR during the upgrade/downgrade.

Reviewer 2: It would not happen as part of an upgrade or downgrade. Pertinent here, in my mind however, is how a cluster that has been upgraded from a version that did not support ingress placement would behave.

This enhancement, to the best of my knowledge, should have no problems
persisting across cluster upgrades. The Test Plan section describes how this
will be tested.

### Version Skew Strategy

Does not apply, to the best of my understanding.

### Operational Aspects of API Extensions

> - The `ControlPlane` value of the `IngressPlacement` parameter value may not be

Reviewer 4: "May" is an unfortunately vague word, so I'm not sure how to interpret this sentence. Does "may not be used" here mean "it is possible that it will not be used" or "is not allowed to be used"?

Author: The latter. Will clarify

used when the cluster's `ControlPlaneTopology` parameter is set to `External`,
this combination is currently considered invalid. This may change in the future.

- This API change only affects the defaulting behavior of the `IngressController` CR,
it does not add any new capabilities to OCP, or give any more flexibility than there

> already was.

Reviewer 3: Given this is only used for defaulting, do we actually need a new field or is there some other way that we could autodetect this and not allow users to configure it at all?

Author: Resolving because me and you are already discussing it in https://github.com/openshift/enhancements/pull/1041#issuecomment-1060638611

- For administrators and support engineers, the `IngressController` is still the
source of truth and where you need to look if you seek to understand the router
placement in practice. Nothing has changed in that regard.

#### Failure Modes

- The `ControlPlane` value of the  `IngressPlacement` parameter value may not be

> used if the cluster's `ControlPlaneTopology` parameter is set to `External`.

Reviewer 3: This should be blocked by a validation of some description

Author: Discussed previously here https://github.com/openshift/enhancements/pull/1041#r817908736

The Ingress Operator would treat such combination as if the `IngressPlacement`
value is actually `Workers` instead, and log a warning to indicate that this
invalid combination has been detected. The Ingress Operator should not fail

> to reconcile `IngressController` CRs due to this invalid combination.

Reviewer 3: This seems like we are hiding an issue with the configuration. I would expect in this case that we would surface the issue, not do any further reconciliation and make the user fix the invalid configuration rather than hiding it and just carrying on with what we think is the right behaviour.

In general, if a user specifies an invalid configuration, we must surface that to them, else they might believe they've set a correct configuration and believe we are doing something that we are in fact not doing

Author: Discussed previously here https://github.com/openshift/enhancements/pull/1041#r817908736

#### Support Procedures

- For administrators and support engineers, the `IngressController` is still the source
of truth and where you need to look for understanding the router placement in practice.
Nothing has changed in that regard.

## Implementation History

Not yet applicable

## Drawbacks

- The pinning of the `IngressController` to the "master" pool is another change
which would make single-node clusters slightly different from multi-node
clusters, and any such difference is naturally not ideal.

- The proposed defaulting behavior for the discussed `IngressController`
parameters is complicated and dependent on three different parameters (infra
topology, control-plane topology, and ingress placement) - such complexity
would probably have to be documented in the CRD definitions and may confuse

> users.

Reviewer 3: It will definitely need to be documented within the CRD definitions. But, this begs the question, if it's so complex, why are we doing it? What is the advantage here?

Is it better to just have a user configure the resources correctly with no defaulting? (I don't think we can actually do that because this API was already released, but it's good to think about what the ideal world would look like and note that)

The complexity here and the apparent magic that will happen behind the scenes, from a UX perspective, worries me. This would need to be very thoroughly documented to be understood, and even then, I think the behaviour could be surprising to users

Reviewer 3: This is useful context, I didn't see it in the enhancement already when I read through, can we make sure to add it

Reviewer 3: This can be resolved, thank you for the updates

once those 2 new control-plane nodes join the cluster). I am not aware of the
other engineering difficulties we would encounter when attempting to support
the expansion of a single-node control-plane into a three-node control-plane,
but I think they would not be trivial.

- Adjust the "baremetal" platform to support single control-plane node
installations and make users and the Assisted Installer (the current popular,
supported method to install single control-plane node clusters) use that
platform instead of the "none" platform. The baremetal platform solves the
issue described in this enhancement with virtual IP addresses/keepalived. This
approach was dismissed due to much higher development efforts and additional
processes that would need to run on the already resource constrained single
control-plane node. Furthermore, even once the baremetal platform is adjusted
to support single-node clusters, the Assisted Installer would have to go
through a lot of development effort in order to make it use the baremetal
platform rather than the "none" platform currently used for single node
installations. This may happen in the future.

> 

Reviewer 4: Another alternative would be to not change the installer at all, but to require users to set the IngressPlacement value before adding a worker.

Author: That would be too late, as the IngressController has already been created at that point with the wrong values. 

Also users could always just fix the IngressController manually (guess that could be listed as an alternative, although it's a bit of "don't enhance anything"), but the idea of this enhancement was to make the situation better with sane defaults so users don't encounter this ingress pitfall.

Reviewer: I had a suggestion along these lines

> * Add both a `nodePlacement` field and a `replicas` field to the ingress config; the ingress operator could directly use these values, if provided, as default values for the corresponding fields in ingresscontrollers.  

https://github.com/openshift/enhancements/pull/996#discussion_r800730204

Maybe worth reconsidering?

Reviewer: Maybe this thread isn't the best place for my comment.  It doesn't solve the fact that something (`ingressPlacement` or `nodePlacement`) needs to be set at install-time in the ingress config.  It's an alternative though to more complicated defaulting logic.

Reviewer 2: If we want to go with `nodePlacement` and `replicas` instead of `ingressPlacement`, I can live with that. Either way, the installer has to make the same decision and have the same domain knowledge.

Reviewer 4: The API alternative proposed here should be documented in the body of the enhancement. I think we say the reason we're not choosing that approach is that it would make the interface with the installer more complex, especially if the configuration API for Ingress changes in the future.

Author: Due to nodePlacement being a rather complex parameter, and it possibly having values other than the ones we're interested in (e.g. [the "linux" label](https://github.com/openshift/cluster-ingress-operator/blob/a6c16b7eefe6d7045f3109cd265c4e891f5791ef/pkg/operator/controller/ingress/deployment.go#L569)), I feel like the installation process dictating the exact values for these fields it's a bit outside of its expertise and a sign that maybe a more general concept like the proposed API is a better fit here

Author: Alternatives section updated to discuss this idea

Reviewer 4: @omertuc I agree, but we should include that in the body of the enhancement so that future readers can see that it was considered.

---
title: single-node-openshift-with-workers
authors:
  - "@omertuc"

> reviewers:

Author: Restoring some comments from the previous PR that got closed:

> romfreiman 2 days ago

@sinnykumari PTAL

_____________________________

> romfreiman 2 days ago

@dhellmann @sdodson who should we add as approvers?
_____________________________
 
 >dhellmann 2 days ago
 
There's an API change here, so we need an API approver to look at it. @mfojtik can you help identify someone with time to review this?

I would expect to see someone from the networking team familiar with ingress, too. @knobunc can you help find a reviewer?
_____________________________
 
> romfreiman 8 hours ago

@tkashem
 _____________________________

> romfreiman 2 days ago

@Miciah who's the approver from network-edge side?

`.spec.replicas` will be chosen according to the value of
`InfrastructureTopology`, namely `1` when `SingleReplica` or `2` when
`HighlyAvailable`. `.spec.nodePlacement` will always just be:

```yaml
nodePlacement:
  nodeSelector:
    matchLabels:
      kubernetes.io/os: linux
      node-role.kubernetes.io/worker: ''
```

However, if the value of `IngressPlacement` is `ControlPlane`, the defaulting
behavior will be different: `.spec.replicas` will be chosen instead according
to the value of `ControlPlaneTopology`; again, `1` when `SingleReplica` or `2`
when `HighlyAvailable`. `.spec.nodePlacement` will be always just be:

```yaml
nodePlacement:
  nodeSelector:
    matchLabels:
      kubernetes.io/os: linux
      node-role.kubernetes.io/master: ''
```

(Note that the `kubernetes.io/os: linux` label is mentioned just because it's
the current behavior, it has no importance in this enhancement)

The installer will detect situations in which it's unlikely the user will want
to set up a load-balancer. For now, those situations include installation of
single control-plane node cluster deployments on "on-prem" platforms such as

> "none" or "vSphere" (although today single control-plane node clusters are only

Reviewer 2: Can we explicitly state which platforms will have this behavior and which will not? Is it **all** on-prem platforms? Also, I think it is a bit confusing to classify "none" as "on-prem". The "none" platform is agnostic and not necessarily on-prem.

Author: Yeah so I guess simply listing all the platforms is the way to go here

Author: The platforms are -
load balancer automatically provisioned / has VIP -
- alibaba
- aws
- azure
- gcp
- ibmcloud
- baremetal

load balancer needed by not provisioned -
- libvirt
- none
- openstack
- ovirt
- vsphere

Is that correct? Did I miss any platform?

Reviewer 2: If this proposal is not applicable to platforms that use VIPs, then the "none" platform is the only platform to which this proposal applies.

For libvirt, the cluster uses local DNS for load balancing.

For openstack, ovirt, and vsphere, the cluster uses VIPs.

Author: Got it, revised.

Reviewer: > I'm personally only really familiar with the installer behavior with cloud (AWS, azure, gcp, where a load balancer is provisioned by default) and the none platform (no load balancer provisioned by the installer by default), so I'm not the best person to answer this. @Miciah said there's also no load-balancer on vsphere...

I don't remember where or when I said that, but it's a little more nuanced as I understand it: vSphere UPI requires the user to set up an external LB; vSphere IPI sets up keepalived-ipfailover with a vip for ingress (and a vip for the API).  

> Yeah so I guess simply listing all the platforms is the way to go here

That's what we do in the IngressController API godoc.

Author: >I don't remember where or when I said that, but it's a little more nuanced as I understand it: vSphere UPI requires the user to set up an external LB; vSphere IPI sets up keepalived-ipfailover with a vip for ingress (and a vip for the API).

@staebler so would that mean we're gonna need more than a simple list of platforms here? A list of (platform,installation method) perhaps? 

Or would it be better to revert to not using an explicit list and to restore the phrasing of "Situations in which the installer provisions keepalived/LB"?

Reviewer 2: The installer does not know the installation method (eg. IPI or UPI) that the user is going to take. As such, the installer cannot use the installer method as a criteria for configuring the Ingress CR manifest.

Ultimately, the installer needs code that will set the ingressPlacement in the Ingress CR manifest. I would prefer that the explicit list of platforms be stipulated here in the enhancement rather than deferring to later when it comes time to write the code to try to figure it out.

Author: >The installer does not know the installation method (eg. IPI or UPI) that the user is going to take. As such, the installer cannot use the installer method as a criteria for configuring the Ingress CR manifest.

In the case of vSphere, could we maybe be looking at `platform.vsphere.apiVIP` / `platform.vsphere.ingressVIP` (see [here](https://docs.openshift.com/container-platform/4.9/installing/installing_vsphere/installing-vsphere-installer-provisioned-customizations.html#installation-installer-provisioned-vsphere-config-yaml_installing-vsphere-installer-provisioned-customizations)) to determine the value of `IngressPlacement`?

Reviewer 4: @omertuc This thread isn't resolved. Is the question answered in the body of the enhancement? If not, do we need an answer?

Author: The enhancement currently only refers to the none-platform, if we're interested in doing this enhancement on other platforms, I'll need the above question answered. This not coming from any real world use requirements, so I guess answering that is not urgent and could potentially be done in a separate enhancement. I wouldn't want to delay this enhancement further over this

> possible on the "none" platform). In those situations, the installer will set

Reviewer 2: Are "none" clusters only possible or only supported?

Author: supported. Will fix

`IngressPlacement` to be `ControlPlane`. Since there's just a single
control-plane node, `ControlPlane` topology would be `SingleReplica` and the
combined effect would be that the `IngressController` will have just a single
replica and be pinned to the single control-plane node. This will then ensure
that the `router-default` deployment created by the Cluster Ingress Operator
will always run on the single control-plane node, and as a result any
`*.apps.<cluster>.<base>` DNS entries which originally pointed at the single
control-plane node will remain correct even in the face of newly added worker
nodes.

In any other situations, the installer will set `IngressPlacement` to
`Workers`, resulting in the same default behavior as before this enhancement,
namely that `IngressController` pod replicas are scheduled on worker nodes and
determined according to the `InfrastructureTopology`.

In the future, when IPI-installed single control-plane node clusters in the
cloud no longer provision a load-balancer by default, they would also benefit
from having the installer set the `IngressPlacement` to `ControlPlane`.

### Risks and Mitigations

> This should make no noticable difference on "regular" single control-plane node

Reviewer 2: What is included in the class of "regular" clusters here?

Author: Single control-plane node clusters with no additional worker nodes, will clarify

```go
type IngressStatus struct {
    // ... existing fields omitted

	// defaultPlacement is set by the installer and lets it control which
	// nodes will host the ingress router pods by default, the options being
	// control-plane nodes or worker nodes.
	//
	// This field works by dictating how the Ingress Operator will set the default
	// values of future IngressController resources' replicas and nodePlacement
	// fields.
	//
	// The value of replicas is set based on the value of a chosen field in the
	// Infrastructure CR. If defaultPlacement was set to ControlPlane, the
	// chosen field will be controlPlaneTopology. If it was set to Workers the
	// chosen field will be infrastructureTopology. Replicas will then be set to 1
	// or 2 with respect to the chosen field's value (which can be either
	// SingleReplica or HighlyAvailable).
	//
	// The value of nodePlacement is adjusted based on defaultPlacement - If
	// defaultPlacement is set to ControlPlane the "node-role.kubernetes.io/worker"
	// label will be added. If defaultPlacement is set to Workers the
	// "node-role.kubernetes.io/master" label will be added.
	//
	// +kubebuilder:validation:Enum:="ControlPlane";"Workers"

> 	// +kubebuilder:default:="Workers"

Reviewer 3: The default value isn't included in auto generated docs, so you may want to add something like below to the godoc string so that the API reference tells users what the default is
```
// When omitted, the default value is Workers.
```

> 	// +kubebuidler:default:="Workers"

Reviewer: ```suggestion
	// +kubebuilder:default:="Workers"
```

	// +optional
	DefaultRouterPlacement DefaultRouterPlacement `json:"defaultRouterPlacement"`
}

> // TopologyMode defines the topology mode of the control/infra nodes.
> type DefaultRouterPlacement string

Reviewer: ```suggestion
// DefaultRouterPlacement defines the default placement of ingress router pods.
type DefaultRouterPlacement string
```

const (
	// "Workers" is for having router pods placed on worker nodes by default
	Workers DefaultPlacement = "Workers"

	// "ControlPlane" is for having router pods placed on control-plane nodes by default
	ControlPlane DefaultPlacement = "ControlPlane"

> )

Reviewer 3: Very much a nit, just thinking about how these will be imported, `ingressv1.Workers`, `ingressv1.ControlPlane`, wondering if you might want to make the names more descriptive, eg, `type DefaultPlacementTopology string` and `WorkersPlacementTopology`/`ControlPlanePlacmentTopology`. I don't have a strong opinion here though and this can be hashed out during the API PR

The sections below go into detail about the field, its possible values, and the
behavior expected from each of them, but in practice following is the proposed
addition to `openshift/api`'s `config/v1/types_ingress.go` file:

```go
type IngressStatus struct {
    // ... existing fields omitted

	// defaultPlacement is set by the installer and lets it control which
	// nodes will host the ingress router pods by default, the options being
	// control-plane nodes or worker nodes.
	//
	// This field works by dictating how the Ingress Operator will set the default
	// values of future IngressController resources' replicas and nodePlacement
	// fields.
	//
	// The value of replicas is set based on the value of a chosen field in the
	// Infrastructure CR. If defaultPlacement was set to ControlPlane, the
	// chosen field will be controlPlaneTopology. If it was set to Workers the
	// chosen field will be infrastructureTopology. Replicas will then be set to 1
	// or 2 with respect to the chosen field's value (which can be either
	// SingleReplica or HighlyAvailable).
	//
	// The value of nodePlacement is adjusted based on defaultPlacement - If

> 	// defaultPlacement is set to ControlPlane the "node-role.kubernetes.io/worker"
> 	// label will be added. If defaultPlacement is set to Workers the
> 	// "node-role.kubernetes.io/master" label will be added.

Reviewer 2: These are backwards, right?
```suggestion
	// defaultPlacement is set to ControlPlane the "node-role.kubernetes.io/master"
	// label will be added. If defaultPlacement is set to Workers the
	// "node-role.kubernetes.io/worker" label will be added.
```

>     //
>     // When omitted, the default value is Workers

Reviewer 2: ```suggestion
	//
	// When omitted, the default value is Workers
```

---
title: single-node-openshift-with-workers
authors:
  - "@omertuc"
reviewers:
  - "@romfreiman"
  - "@eranco74"
  - "@tsorya"
  - "@dhellmann"
  - "@Miciah"
  - "@bparees"

>   - "@JoelSpeed"

Reviewer 4: ```suggestion
  - "@JoelSpeed" - API
```

>   - "@staebler"

Reviewer 4: ```suggestion
  - "@staebler" - installer
```

  - "@derekwaynecarr"
approvers:

>   - TBD

Reviewer 5: The enhancement is open for almost 3 months already. We're about to miss the deadline. If you do see some critical points, please raise them asap. Otherwise, I really would like to merge it till the end of the week

api-approvers: # in case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers)

>   - TBD

Reviewer 4: ```suggestion
  - "@JoelSpeed"
```

issue itself.

One of the benefits of installing single-node clusters is the simplicity of not
having to deal with load-balancing and virtual IPs, as these don't provide much
value when there's only a single node behind them.

As a result, current common ways of installing single control-plane node
clusters today (mainly the Assisted Installer) avoid the usage of load
balancers or virtual IPs for API and ingress. There has also been some recent
effort to determine how single control-plane node cluster deployments on clouds
may be adjusted in order to reduce their costs, and one of the conclusions is
that getting rid of the load balancer installed by default by the IPI installer
results in major cost savings.

A user installing a single control-plane node cluster on "none"-platform will
be tempted to simply point their DNS entries directly at the IP address of the
single node that they just installed.

Similarly, in the Assisted Installer, the user is able to complete the
installation without needing to define any DNS entries. This is currently made
possible by injecting a `MachineConfig` manifest targeting the "master" pool

> containing configuration for a dnsmasq server. The dnsmasq server is configured

Reviewer 6: I wasn't aware of that; seems like a nontrivial bit.  Where is that code?

Reviewer 5: In assisted installer. We just inject a manifest via standard installer apis

Reviewer 4: @cgwalters My impression is that this is really background info, and doesn't affect the change proposed here. Is that right?

Author: Not sure whether that question is directed towards me or not, but yeah, I was just providing it as background information

Reviewer 4: The question was for @cgwalters. I don't think there's any impact, but want to make sure before we merge this.

Reviewer 6: No, not a blocking concern.

Reviewer 4: Thanks, Colin

adjust how it behaves in the absence of these parameters. The exact details of
this behavior are described below.

The goal of this new `DefaultPlacement` field is to give the installer a way to
affect the general placement of the `IngressController`. This is regardless of
whether an `IngressController` is created during installation with the
`.spec.replicas` and `.spec.nodePlacement` omitted or whether no
`IngressController` was created during an installation at all. Today, when no
`IngressController` is created during installation, the Cluster Ingress Operator
creates a default `IngressController`, in this case we can go one of two ways -

- The Cluster Ingress Operator leaves those fields empty in the newly created
`IngressController`

- The cluster Ingress Operator populates these fields in the newly created
`IngressController`

Choosing one way or the other is arbitrary and shouldn't matter due to the
immutable nature of this new field - the eventual `Deployment` that gets
created would be identical in any case. I personally prefer the second option
since it makes the fields explicit in the `IngressController` itself and not

> just in `Deployments` that get created.

Reviewer 3: I think there are differences, if you leave the fields empty, the controller can change the values at runtime later, if you have the controller populate the values (this should be via admission webhook), then the values are now set and can not be changed later unless the user does so. I was under the impression we wanted to go with the first option of leaving them empty for now to maintain existing behaviour?

Author: Note that what I'm discussing here is specifically what the controller would do to the IngressController it **creates itself**, as can be seen in [this](https://github.com/openshift/cluster-ingress-operator/blob/0f36470250c3089769867ebd72e25c413a29cda2/pkg/operator/operator.go#L218-L254) code - it's a `ensureDefaultIngressController` function that creates an IngressController if one was not created during installation - whether it bakes the replicas/placement into that particular IngressController or not is of no consequence, due to the immutability of the new defaultPlacement parameter

The sections below go into detail about the field, its possible values, and the
behavior expected from each of them, but in practice following is the proposed
addition to `openshift/api`'s `config/v1/types_ingress.go` file:

```go
type IngressStatus struct {
    // ... existing fields omitted

	// defaultPlacement is set at installation time to control which
	// nodes will host the ingress router pods by default. The options are
	// control-plane nodes or worker nodes.
	//
	// This field works by dictating how the Cluster Ingress Operator will 
    // consider unset replicas and nodePlacement fields in IngressController
    // resources when creating the corresponding Deployments.
    //
    // See the documentation for the IngressController replicas and nodePlacement

>     // fields for more information.

Reviewer 3: Please fix indentation here and on other fields in the go types

	// cluster config.openshift.io/v1/ingresses status. 
	//
	// The value of replicas is set based on the value of a chosen field in the
	// Infrastructure CR. If defaultPlacement is set to ControlPlane, the
	// chosen field will be controlPlaneTopology. If it is set to Workers the
	// chosen field will be infrastructureTopology. Replicas will then be set to 1
	// or 2 based whether the chosen field's value is SingleReplica or
	// HighlyAvailable, respectively.
	//
	// +optional
	Replicas *int32 `json:"replicas,omitempty"`

	// ... existing fields omitted
}

type NodePlacement struct {
	// ... existing fields omitted

	// nodeSelector is the node selector applied to ingress controller
	// deployments.
	//
	// If set, the specified selector is used and replaces the default.
	// 
	// If unset, the default depends on the value of the defaultPlacement
    // field in the cluster config.openshift.io/v1/ingresses status.
    // 
    // When defaultPlacement is Workers, the default is:
	//
	//   kubernetes.io/os: linux
	//   node-role.kubernetes.io/worker: ''
    //
    // When defaultPlacement is ControlPlane, the default is:
    //
    //   kubernetes.io/os: linux

>     //   node-role.kubernetes.io/master: ''

Reviewer 3: NIt: indentation still off here

target.

I do not believe this enhancement has any security implications.

## Design Details

### Open Questions

None that I can think of at the moment

### Test Plan

- Add unit tests to the Cluster Ingress Operator to make sure the
IngressController resource is generated as exepected.

- Add periodic nightly tests which install single-node in the cloud, add a few
worker nodes to it, then run conformance tests to make sure we don't run into
any problems not described in this enhancement.

- Add periodic nightly tests which install a single-node "none"-platform cluster,
add worker nodes to it, and check that ingress traffic still works as expected
and recovers even after the `router-default` pod gets deleted and rescheduled.
Make sure this is still true even after upgrades.

- Add tests on both cloud / "none"-platform that check that a single-node cluster
with additional workers recovers after the single control-plane node reboots by
running conformance tests post-reboot.

- Add tests on both cloud / "none"-platform that check that a single-node cluster
with additional workers recovers after an upgrade by running conformance tests
post-upgrade.

> TODO: Describe day-1-workers tests?

Reviewer 4: @omertuc I think this is covered by the item on line 444? Let me know if you're planning to add more details.

### Graduation Criteria

**Note:** *Section not required until targeted at a release.*

TBD

#### Dev Preview -> Tech Preview

TBD

#### Tech Preview -> GA

TBD

#### Removing a deprecated feature

N/A

### Upgrade / Downgrade Strategy

The new `DefaultPlacement` API field will have a default value of `Workers`.
The behavior of the Cluster Ingress Operator when the `DefaultPlacement` field
has the value `Workers` should be identical to its current behavior (before
this enhancement). This means that clusters that go through an upgrade from a
version before this enhancement to a later version which includes this
enhancement will maintain their current behavior, this enhancement should not
change anything for them.

> If the value of `DefaultPlacement` is empty (TODO: Could this possibly happen

Reviewer 4: @omertuc This TODO item looks like an open question. Is it answered elsewhere?

Author: It is not answered elsewhere, however, my understanding of how CRDs work has improved and I'm pretty certain of the answer, so I'll remove the TODO

if the the Cluster Ingress Operator reads a resource created according to the
old Ingress CRD that didn't have this value? Not sure how defaults work in this
scenario, in any case, it's not crucial) the Cluster Ingress Operator should
make sure to treat it as if it were to have the value `Workers`.

### Version Skew Strategy

Does not apply, to the best of my understanding.

### Operational Aspects of API Extensions

- Since the `DefaultPlacement` field is part of the status of the
Ingress config CR, it is clear that users should not modify this field.

- This API change only affects the defaulting behavior of the `IngressController`
CR, it does not add any new capabilities to OCP, or give any more flexibility
than there already was.

- For administrators and support engineers, the `IngressController` is still the
source of truth and where you need to look if you seek to understand the router
placement in practice. However, now the `replicas` and `nodePlacement` on
`IngressController` may be empty, so administrators and support engineers 
can refer to the corresponding Deployment that got created if they wish to

> know the practical values.

Reviewer 4: Why wouldn't we have the IngressController fill in the settings it is using a default, so the user doesn't have to guess?

Author: Do you mean the ingress operator should actively reconcile all IngressController CRs all the time to populate these values when unset with the defaults that would otherwise be set in the corresponding Deployment?

I guess it could do that, I'm not the one to say how typical it is for an operator to do something like that. I guess this is something we can can discuss in the ingress PR? I don't have an opinion here

Reviewer 4: Yes, that's what I meant.

@JoelSpeed @Miciah is this worth resolving here?

Author: (Added 9ef9aac54541c72b764019dd756b7c7f725f6980 to mention this idea)

Reviewer 3: @omertuc and I discussed this on a call before, I thought there was a note already. Upshot of our conversation is that there are two ways you can default, either via webhook or via an operator at runtime. For this particular field, if we default via the operator at runtime, that allows us flexibility to change the defaults in the future (note the comment on the API about this). If we default via a webhook, whatever gets set on the resource is now there and set, we cannot change that default later because there will be existing users with the default encoded in their resource even though they had no opinion on it.

For this particular example, we agreed a runtime defaulting in the controller was likely the best approach. Adding a webhook defaulting option to the alternatives is a reasonable ask though IMO (where we can document why we aren't doing that)

I'm going to add a section talking about defaults to the API conventions doc I'm working on.

Author: @JoelSpeed my understanding is that what's being suggested in this thread is neither what you would call "webhook defaulting" nor "operator at runtime defaulting" - it's more like the operator continuously doing the defaulting work the webhook would otherwise do, by making sure those values are not left unset in the IngressControllers, but are instead set to their default values.

Author: That's right, one difference being that it overcomes the problem of the webhook not being present during installation time

Reviewer 4: Yes, Omer's description of what I was asking about is correct.

If we want the default to be able to change, then leaving the values unset is the right way to tell the difference between values the user gives and the defaults. We could reflect the actual values being used in the status field to aid in debugging, but if those fields aren't already there then that's probably a different enhancement.

Omer, could you add the conclusion from this thread to the body of the doc and then resolve it, please?

Author: My conclusion remains what I've added in 9ef9aac54541c72b764019dd756b7c7f725f6980, do you think I missed any important detail?

Reviewer 4: I think what you've written covers it adequately. Thanks!
